### PR TITLE
Split LMBert model in two

### PR DESCRIPTION
**Pull request by sgugger**
As discussed in #4711, the `BertForMaskedLM` model should be split in two to avoid having two different labels argument, one model for causal LM, one for masked LM. This PR follows up on that and does the split.

It introduces a new `BertLMHeadModel` (also added to the `__init__` and the docs) with a test. As discussed, there is no deprecation warning if someone tries to use the `lm_labels` in `BertForMaskedLM` (since it was experimental), but an error message telling the user to use `BertLMHeadModel`.

I did not add `BertLMHeadModel` in the automodel logic since we probably want users to use causal models for this? Let me know if I should add it even if it's not the best model for that task.

I also removed `lm_labels` in the `EncoderDecoderModel` since it was only there to support that argument in `BertForMaskedLM` (which then removes the corresponding test).

### CODE REVIEW

File: src/transformers/modeling_bert.py [hunk 1]
```diff
@@ -935,6 +1048,7 @@ def forward(
                 DeprecationWarning,
             )
             labels = kwargs.pop("masked_lm_labels")
+        assert "lm_labels" not in kwargs, "Use `BertWithLMHead` for autoregressive language modeling task."
```
Comment (LysandreJik) [inline src/transformers/modeling_bert.py:1051]: It wouldn't be in the kwargs as it's still in the method signature

**Review by LysandreJik — COMMENTED** [review]
I think this is cool, it simplifies the `BertForMaskedLM` model and makes the API more coherent.

However, this is a breaking change. The model now behaves differently, with part of its functions being removed.

I'm not sure that breaking change is worth it.

File: src/transformers/modeling_bert.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:1049]: Good point, forgot to remove that arg from the signature.

**Comment by sgugger** [comment]
It's possible to do it in a non-breaking way with a deprecation warning if a not-None `labels_lm` is passed to `BertForMaskedLM`. I was following the discussion of #4711 that implied it was okay to have a breaking change for this.

**Comment by patrickvonplaten** [comment]
I'm fine with this PR. IMO, `BertForMaskedLM` was never really used before for causal language modeling except when using Bert in an encoder-decoder setting and the encoder-decoder code is not really released yet. Also since we keep the same names for the submodules `self.bert` and `self.cls`, there won't be any errors or inconsistencies when loading pretrained weights into the Bert2Bert encoder-decoder. 

In my opinion, this change is actually quite important to have a clean separation between masked lm and causal lm (Reformer and Longformer will eventually run into the same issue).

The heavily used `BertForMaskedLM` for the normal masked encoder bert model does not change at all except for `lm_labels`, so that's good IMO as well. 

One thing which is problematic though is that the `MODEL_WITH_LM_HEAD_MAPPING` contains a mixture of causal models and masked encoder models: https://github.com/huggingface/transformers/blob/02e5f79662d72cccdca81a47e3001a5f6d36e5b1/src/transformers/modeling_auto.py#L187

Now since Bert has both a causal model and a masked encoder model we need two mappings. 

I would suggest here to create 2 new mappings `MODEL_FOR_MASKED_LM_MAPPING` and `MODEL_FOR_CAUSAL_LM_MAPPING` and two new AutoModels: `AutoModelForMaksedLM` , `AutoModelForCausalLM` for now keep `AutoModelWithLMHead` as it is and add a depreciated warning to it. We can add `BertLMHeadModel` to `MODEL_FOR_CAUSAL_LM_MAPPING` and change to `AutoModelForCausalLM` in the encoder-decoder file.

File: src/transformers/modeling_bert.py [hunk 2]
```diff
@@ -852,9 +852,10 @@ def forward(
         return outputs  # (loss), prediction_scores, seq_relationship_score, (hidden_states), (attentions)
 
 
-# TODO: Split with a different BertWithLMHead to get rid of `lm_labels` here and in encoder_decoder.
-@add_start_docstrings("""Bert Model with a `language modeling` head on top. """, BERT_START_DOCSTRING)
-class BertForMaskedLM(BertPreTrainedModel):
+@add_start_docstrings(
+    """Bert Model with a `language modeling` head on top for CLM fine-tuning. """, BERT_START_DOCSTRING
+)
+class BertLMHeadModel(BertPreTrainedModel):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:858]: I think the init should have an assert statement now to make sure `config.is_decoder=True` or even set `config.is_decoder=True` with a logging info that it does so. Using `BertLMHeadModel` without a causal mask does not make much sense IMO. 

Also, currently `BertLayer` always initializes the cross-attention layer if `config.is_decoder=True`, but doesn't use it if no `encoder_hidden_states` are passed to forward => so there is not really a problem atm when using the model on its own (not in an encoder-decoder framework). BUT, this should be corrected with an additional check `self.config.is_encoder_decoder`  before adding the cross-attention layers.

File: src/transformers/modeling_bert.py [hunk 3]
```diff
@@ -953,26 +1065,12 @@ def forward(
...
-            prediction_scores = prediction_scores[:, :-1, :].contiguous()
-            lm_labels = lm_labels[:, 1:].contiguous()
-            loss_fct = CrossEntropyLoss()
-            ltr_lm_loss = loss_fct(prediction_scores.view(-1, self.config.vocab_size), lm_labels.view(-1))
-            outputs = (ltr_lm_loss,) + outputs
-
-        return outputs  # (ltr_lm_loss), (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
+        return outputs  # (masked_lm_loss), prediction_scores, (hidden_states), (attentions)
 
     def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:1075]: The `prepare_inputs_for_generation` function has to be added to `BertLMHeadModel` so that we can generate with Bert2Bert encoder-decoder. 

We can now split the function into two cases: 1) self.config.is_decoder is False 2) self.config.is_decoder is True and add the function to its corresponding models.

File: src/transformers/modeling_bert.py [hunk 4]
```diff
@@ -878,17 +879,129 @@ def forward(
...
+        return {"input_ids": input_ids, "attention_mask": attention_mask}
+
+
+@add_start_docstrings("""Bert Model with a `language modeling` head on top. """, BERT_START_DOCSTRING)
+class BertForMaskedLM(BertPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.bert = BertModel(config)
+        self.cls = BertOnlyMLMHead(config)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:981]: I think we could also do a check here that `self.is_decoder == False`.

File: src/transformers/modeling_bert.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:1075]: It is in `BertLMHeadModel`. Do you mean only the first test should be in `BertLMHeadModel` and the only the second test should be in `BertForMaskedLM`?

Comment (patrickvonplaten) [reply src/transformers/modeling_bert.py:1075]: Sorry my bad! You did add the function. 

Yeah, I think now we can remove some logic from the `prepare_inputs_for_generation` since `config.is_decoder` should not be true anymore for `BertForMaskedLM` and `config.is_decoder` should always be True for `BertLMHeadModel` (since it should use the causal mask` for language generation). But maybe this could be done in a separate PR.

File: tests/test_modeling_encoder_decoder.py [hunk 5]
```diff
@@ -288,38 +287,6 @@ def create_and_check_bert_encoder_decoder_model_labels(
...
-        attention_mask,
-        encoder_hidden_states,
-        decoder_config,
-        decoder_input_ids,
-        decoder_attention_mask,
-        lm_labels,
-        **kwargs
-    ):
-        encoder_model = BertModel(config)
-        decoder_model = BertForMaskedLM(decoder_config)
```
Comment (patrickvonplaten) [inline tests/test_modeling_encoder_decoder.py:304]: Can we change to `BertLMHeadModel` in all tests here.

File: src/transformers/modeling_bert.py [hunk 6]
```diff
@@ -878,17 +879,129 @@ def forward(
...
+            attention_mask = torch.cat(
+                [attention_mask, attention_mask.new_zeros((attention_mask.shape[0], 1))], dim=-1
+            )
+
+            dummy_token = torch.full(
+                (effective_batch_size, 1), self.config.pad_token_id, dtype=torch.long, device=input_ids.device
+            )
+            input_ids = torch.cat([input_ids, dummy_token], dim=1)
+
+        return {"input_ids": input_ids, "attention_mask": attention_mask}
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bert.py:972]:
```python 
    def prepare_inputs_for_generation(self, input_ids, attention_mask=None, **model_kwargs):

        # if model is used as a decoder in encoder-decoder model, the decoder attention mask is created on the fly
        if attention_mask is None:
            attention_mask = input_ids.new_ones(input_ids.shape)

        return {"input_ids": input_ids, "attention_mask": attention_mask}
```

should be enough here

File: src/transformers/modeling_bert.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_bert.py:1075]: Here, this should be enough:

```python 
        effective_batch_size = input_ids.shape[0]

        #  add a dummy token
        assert self.config.pad_token_id is not None, "The PAD token should be defined for generation"
        attention_mask = torch.cat(
            [attention_mask, attention_mask.new_zeros((attention_mask.shape[0], 1))], dim=-1
        )
        dummy_token = torch.full(
            (effective_batch_size, 1), self.config.pad_token_id, dtype=torch.long, device=input_ids.device
        )
        input_ids = torch.cat([input_ids, dummy_token], dim=1)

        return {"input_ids": input_ids, "attention_mask": attention_mask}
```

Comment (sgugger) [reply src/transformers/modeling_bert.py:1064]: Done.

File: tests/test_modeling_encoder_decoder.py (hunk 5 above)
Comment (sgugger) [reply tests/test_modeling_encoder_decoder.py:304]: Done.
Note that the tests using automodel pick the wrong decoder for now.

File: src/transformers/modeling_bert.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:970]: We can't raise an exception here before the automodel in encoder_decode picks the right LM model.

File: src/transformers/modeling_bert.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:858]: Added an assert.

File: src/transformers/modeling_bert.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/modeling_bert.py:961]: Done.

**Comment by sgugger** [comment]
I agree with @patrickvonplaten on the need to split `AutoModelWithLMHead` in two. Note that if the name `AutoModelForCausalLM` is picked, we should then rename (with a deprecation first of course) all `ModeltypeLMHeadModel` to `ModeltypeForCausalLM` for consistency (and clarity since just saying it has an LM head doesn't tell us if it's intended to be masked or causal).

**Comment by LysandreJik** [comment]
I agree that having two additional `AutoXXX` classes for the distinction between masked/causal would be nice. We should, however, keep the `AutoModelWithLMHead` class available for backwards compatibility.

However, I don't agree with renaming all causal model with language modeling heads `XXXForCausalLM`. It would be more consistent, but is an aesthetic change with a very big breaking change. Even adding aliases to keep backwards compatibility would create a large overhead for the user, in my opinion, as all those classes would exist twice when importing from the library.

**Comment by sgugger** [comment]
In that case I would advocate to keep `AutoModelWithLMHead` for causal language models and only add an `AutoModelForMaskedLM`. Consistency is cosmetic, I agree, but it also helps not confusing beginners.

File: src/transformers/modeling_bert.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/modeling_bert.py:969]: Ok, we could do this in a separate PR

**Comment by patrickvonplaten** [comment]
1) For now, I think the best solution would be to keep `AutoModelForMaskedLM` as it is and add two new `AutoXXX` classes. The EncoderDecoderModel would be the first model to use  `AutoModelForCausalLM` in its code. 

`AutoModelWithLMHead` is heavily used for all kinds of masked bert encoder models, so if we create an `AutoModelForMaskedLM` and move `BertForMaskedLM` there, we would have a lot of breaking change. I think we could add a depreciation warning to `AutoModelWithLMHead` though.  

2) I'm a bit indifferent to renaming all other model classes. While I'm also a big fan of consistency I agree with @LysandreJik in that I think it's a big user-facing API change that is not really urgent atm.

**Comment by julien-c** [comment]
In the short term, I would advocate only exposing the classical "masked-lm" flavour of BERT through AutoModelWithLMHead (as is done in this PR), and not even documenting/adding BertLMHeadModel to the `__init__`, as it's only used as a building block to other models.

In the longer term, I'd be ok with creating `AutoModelFor{Masked,Causal}LM` (name TBD for the second one) and not even creating a deprecation for `AutoModelWithLMHead`, forcing users to explicitly choose one or the other. This would need to be a major release though.

**Comment by LysandreJik** [comment]
@julien-c as long as we do a major release for this, I'm all for this!

**Review by sshleifer — APPROVED** [review]
read this, very nice!

**Comment by patrickvonplaten** [comment]
> In the short term, I would advocate only exposing the classical "masked-lm" flavour of BERT through AutoModelWithLMHead (as is done in this PR), and not even documenting/adding BertLMHeadModel to the `__init__`, as it's only used as a building block to other models.
> 
> In the longer term, I'd be ok with creating `AutoModelFor{Masked,Causal}LM` (name TBD for the second one) and not even creating a deprecation for `AutoModelWithLMHead`, forcing users to explicitly choose one or the other. This would need to be a major release though.

For the encoder decoder models, I think we need `BertLMHeadModel` in the `init` and we would also need a `AutoModelForCausalLM`. Here: https://github.com/huggingface/transformers/blob/29c36e9f3678702e5ffd3fe2f1c9f6c1d6672578/src/transformers/modeling_encoder_decoder.py#L160 we need to instantiate a `BertWithCausalLM`

**Review by thomwolf — APPROVED** [review]

**Comment by thomwolf** [comment]
I'm fine either way, I think you guys got all the important issues (backward compatibility versus cleanly building the future). I like what @patrickvonplaten and @julien-c are proposing.

**Review by LysandreJik — APPROVED** [review]
Let's not forget to write an explanation of why that breaking change was necessary in the release notes.

**Comment by patrickvonplaten** [comment]
This currently breaks the encoder-decoder framework `from_encoder_decoder_pretrained()` method. Will do a PR tomorrow to fix it.
